File: implementations/go/examples/internal/domain/order/order.go
package order

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Status string

const (
	StatusPending   Status = "pending"
	StatusPaid      Status = "paid"
	StatusCancelled Status = "cancelled"
)

type ItemInput struct {
	ItemID   int
	Name     string
	Price    int
	Quantity int
}

type Order struct {
	OrderID  string
	UserID   string
	Items    []Item
	Status   Status
	events   []OrderCancelled
}

func New(userID string, inputs []ItemInput) (*Order, error) {
	if len(inputs) == 0 {
		return nil, ErrEmptyItems
	}
	items := make([]Item, 0, len(inputs))
	for _, in := range inputs {
		item, err := newItem(in.ItemID, in.Name, in.Price, in.Quantity)
		if err != nil {
			return nil, fmt.Errorf("invalid item %d: %w", in.ItemID, err)
		}
		items = append(items, item)
	}
	return &Order{
		OrderID: uuid.NewString(),
		UserID:  userID,
		Items:   items,
		Status:  StatusPending,
	}, nil
}

func Reconstitute(orderID, userID string, items []Item, status Status) *Order {
	return &Order{OrderID: orderID, UserID: userID, Items: items, Status: status}
}

func (o *Order) Cancel(reason string) error {
	if o.Status == StatusCancelled {
		return ErrAlreadyCancelled
	}
	if o.Status == StatusPaid {
		return ErrPaidNotCancellable
	}
	o.Status = StatusCancelled
	o.events = append(o.events, OrderCancelled{
		OrderID:     o.OrderID,
		Reason:      reason,
		CancelledAt: time.Now(),
	})
	return nil
}

func (o *Order) TotalAmount() int {
	total := 0
	for _, item := range o.Items {
		total += item.Price * item.Quantity
	}
	return total
}

func (o *Order) DomainEvents() []OrderCancelled { return o.events }
func (o *Order) ClearEvents()                   { o.events = nil }

File: implementations/fastapi/examples/src/order/domain/order.py
from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Literal

from .errors import OrderAlreadyCancelledError, OrderEmptyItemsError, OrderPaidNotCancellableError
from .order_cancelled import OrderCancelled
from .order_item import OrderItem

OrderStatus = Literal["pending", "paid", "cancelled"]


@dataclass
class ItemInput:
    item_id: int
    name: str
    price: int
    quantity: int


class Order:
    def __init__(
        self,
        order_id: str,
        user_id: str,
        items: list[OrderItem],
        status: OrderStatus,
    ) -> None:
        self.order_id = order_id
        self.user_id = user_id
        self.items = items
        self.status: OrderStatus = status
        self._events: list[OrderCancelled] = []

    @classmethod
    def create(cls, user_id: str, inputs: list[ItemInput]) -> Order:
        if not inputs:
            raise OrderEmptyItemsError()
        items = [OrderItem(i.item_id, i.name, i.price, i.quantity) for i in inputs]
        return cls(
            order_id=str(uuid.uuid4()),
            user_id=user_id,
            items=items,
            status="pending",
        )

    def cancel(self, reason: str) -> None:
        if self.status == "cancelled":
            raise OrderAlreadyCancelledError()
        if self.status == "paid":
            raise OrderPaidNotCancellableError()
        self.status = "cancelled"
        self._events.append(OrderCancelled(
            order_id=self.order_id,
            reason=reason,
            cancelled_at=datetime.utcnow(),
        ))

    def total_amount(self) -> int:
        return sum(item.price * item.quantity for item in self.items)

    def pull_events(self) -> list[OrderCancelled]:
        events, self._events = self._events, []
        return events
